test/pylib/scylla_cluster.py

```diff
@@ -534,19 +525,27 @@ def _seeds(self) -> List[str]:
 
     async def add_server(self) -> ServerInfo:
         """Add a new server to the cluster"""
+        self.is_dirty = True
+        # Scylla assumes all instances of a cluster use the same port,
+        # so each instance needs an own IP address.
+        ip_addr = IPAddress(await self.host_registry.lease_host())
+        seeds = self._seeds()
+        if not seeds:
+            seeds = [ip_addr]
         params = ScyllaCluster.CreateServerParams(
             cluster_name = self.name,
-            seeds = self._seeds(),
+            ip_addr = ip_addr,
```

Comment:
any step after leasing and till the server is added to the list may fail and throw. At this point the IP address will be leaked. Please maintain an independent registry of leased IP addresses. the server may be expected to throw at start by the test, which invokes it as a test from the rest api, at no point we should leak external resources.

Reply:
It would be also reasonable perhaps to lease all IP addresses up-front, before we use them to create servers one-by-one. It's best to wait on a lock without holding resources.

Reply:
> any step after leasing and till the server is added to the list may fail and throw

I don't think it's reasonable to assume that e.g. creating the `CreateServerParams` tuple throws. If it does, that's probably OOM or some other fatal internal Python error, and the entire thing will be killed anyway. Similarly, I think it's reasonable to assume that `self._seeds()` doesn't throw. And beyond that, observe that there's a `try` block, and we release the host in `except` block.

Reply:
> Please maintain an independent registry of leased IP addresses.

In fact I do maintain such a registry later, it's necessary when we start using the same IP for multiple servers (which can happen during replace).